// livekit-server-sdk/dist/proto/livekit_webhook.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookEvent = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const _m0 = __importStar(require("protobufjs/minimal"));
const livekit_models_1 = require("./livekit_models");
const livekit_recording_1 = require("./livekit_recording");
const livekit_egress_1 = require("./livekit_egress");
exports.protobufPackage = "livekit";
function createBaseWebhookEvent() {
    return {
        event: "",
        room: undefined,
        participant: undefined,
        recordingInfo: undefined,
        egressInfo: undefined,
        track: undefined,
        id: "",
        createdAt: 0,
    };
}
exports.WebhookEvent = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.event !== "") {
            writer.uint32(10).string(message.event);
        }
        if (message.room !== undefined) {
            livekit_models_1.Room.encode(message.room, writer.uint32(18).fork()).ldelim();
        }
        if (message.participant !== undefined) {
            livekit_models_1.ParticipantInfo.encode(message.participant, writer.uint32(26).fork()).ldelim();
        }
        if (message.recordingInfo !== undefined) {
            livekit_recording_1.RecordingInfo.encode(message.recordingInfo, writer.uint32(42).fork()).ldelim();
        }
        if (message.egressInfo !== undefined) {
            livekit_egress_1.EgressInfo.encode(message.egressInfo, writer.uint32(74).fork()).ldelim();
        }
        if (message.track !== undefined) {
            livekit_models_1.TrackInfo.encode(message.track, writer.uint32(66).fork()).ldelim();
        }
        if (message.id !== "") {
            writer.uint32(50).string(message.id);
        }
        if (message.createdAt !== 0) {
            writer.uint32(56).int64(message.createdAt);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWebhookEvent();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.event = reader.string();
                    break;
                case 2:
                    message.room = livekit_models_1.Room.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.participant = livekit_models_1.ParticipantInfo.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.recordingInfo = livekit_recording_1.RecordingInfo.decode(reader, reader.uint32());
                    break;
                case 9:
                    message.egressInfo = livekit_egress_1.EgressInfo.decode(reader, reader.uint32());
                    break;
                case 8:
                    message.track = livekit_models_1.TrackInfo.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.id = reader.string();
                    break;
                case 7:
                    message.createdAt = longToNumber(reader.int64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            event: isSet(object.event) ? String(object.event) : "",
            room: isSet(object.room) ? livekit_models_1.Room.fromJSON(object.room) : undefined,
            participant: isSet(object.participant)
                ? livekit_models_1.ParticipantInfo.fromJSON(object.participant)
                : undefined,
            recordingInfo: isSet(object.recordingInfo)
                ? livekit_recording_1.RecordingInfo.fromJSON(object.recordingInfo)
                : undefined,
            egressInfo: isSet(object.egressInfo)
                ? livekit_egress_1.EgressInfo.fromJSON(object.egressInfo)
                : undefined,
            track: isSet(object.track) ? livekit_models_1.TrackInfo.fromJSON(object.track) : undefined,
            id: isSet(object.id) ? String(object.id) : "",
            createdAt: isSet(object.createdAt) ? Number(object.createdAt) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        message.event !== undefined && (obj.event = message.event);
        message.room !== undefined &&
            (obj.room = message.room ? livekit_models_1.Room.toJSON(message.room) : undefined);
        message.participant !== undefined &&
            (obj.participant = message.participant
                ? livekit_models_1.ParticipantInfo.toJSON(message.participant)
                : undefined);
        message.recordingInfo !== undefined &&
            (obj.recordingInfo = message.recordingInfo
                ? livekit_recording_1.RecordingInfo.toJSON(message.recordingInfo)
                : undefined);
        message.egressInfo !== undefined &&
            (obj.egressInfo = message.egressInfo
                ? livekit_egress_1.EgressInfo.toJSON(message.egressInfo)
                : undefined);
        message.track !== undefined &&
            (obj.track = message.track ? livekit_models_1.TrackInfo.toJSON(message.track) : undefined);
        message.id !== undefined && (obj.id = message.id);
        message.createdAt !== undefined &&
            (obj.createdAt = Math.round(message.createdAt));
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseWebhookEvent();
        message.event = (_a = object.event) !== null && _a !== void 0 ? _a : "";
        message.room =
            object.room !== undefined && object.room !== null
                ? livekit_models_1.Room.fromPartial(object.room)
                : undefined;
        message.participant =
            object.participant !== undefined && object.participant !== null
                ? livekit_models_1.ParticipantInfo.fromPartial(object.participant)
                : undefined;
        message.recordingInfo =
            object.recordingInfo !== undefined && object.recordingInfo !== null
                ? livekit_recording_1.RecordingInfo.fromPartial(object.recordingInfo)
                : undefined;
        message.egressInfo =
            object.egressInfo !== undefined && object.egressInfo !== null
                ? livekit_egress_1.EgressInfo.fromPartial(object.egressInfo)
                : undefined;
        message.track =
            object.track !== undefined && object.track !== null
                ? livekit_models_1.TrackInfo.fromPartial(object.track)
                : undefined;
        message.id = (_b = object.id) !== null && _b !== void 0 ? _b : "";
        message.createdAt = (_c = object.createdAt) !== null && _c !== void 0 ? _c : 0;
        return message;
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
function longToNumber(long) {
    if (long.gt(Number.MAX_SAFE_INTEGER)) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    return long.toNumber();
}
if (_m0.util.Long !== long_1.default) {
    _m0.util.Long = long_1.default;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
